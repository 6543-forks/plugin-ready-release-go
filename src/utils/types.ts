import { ExecFunction } from "shelljs";
import { Forge } from "../forges/forge";
import { Config } from "./config";
import { Git } from "./git";

export type PromiseOrValue<T> = Promise<T> | T;

export type Exec = ExecFunction;

export type CommandContext = {
  exec: Exec;
  config: Config;
  forge: Forge;
  git: Git;
};

export type HookContext = {
  exec: Exec;
  nextVersion: string;
};

export type UserConfig = {
  /**
   * Get the next version to release.
   */
  getNextVersion: (ctx: HookContext) => PromiseOrValue<string>;

  /**
   * Get the branch the release will be created from.
   *
   * Normally main or master
   */
  getReleaseBranch: (ctx: HookContext) => PromiseOrValue<string>;

  /**
   * Get the release description.
   *
   * Used as pull-request and release description.
   * By default it's `Release of ${version}`
   */
  getReleaseDescription: (ctx: HookContext) => PromiseOrValue<string>;

  /**
   * Get the branch used for the release pull request
   *
   * By default it's `next-release/${version}`
   */
  getPullRequestBranch: (o: { version: string }) => PromiseOrValue<string>;

  /**
   * Run before the release pull request is created.
   *
   * Could be used to update changelog, version, etc
   */
  beforePrepare: (ctx: HookContext) => PromiseOrValue<boolean | void>;

  /**
   * Run after the release pull request was created.
   *
   * Could be used to notify maintainers about review, etc
   */
  afterPrepare: (ctx: HookContext) => PromiseOrValue<boolean | void>;

  /**
   * Run before the release is created.
   *
   * Could be used to create release notes, etc
   */
  beforeRelease: (ctx: HookContext) => PromiseOrValue<boolean | void>;

  /**
   * Run after the release was created.
   *
   * Could be used to notify maintainers about release, publish release artifacts, etc
   */
  afterRelease: (ctx: HookContext) => PromiseOrValue<boolean | void>;
};

export const defineConfig = (config: Partial<UserConfig>) => config;
