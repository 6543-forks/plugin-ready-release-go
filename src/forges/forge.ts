export abstract class Forge {
  abstract createOrUpdatePullRequest(options: {
    draft: boolean;
    title: string;
    description: string;
    owner: string;
    repo: string;
    sourceBranch: string;
    targetBranch: string;
  }): Promise<void>;

  abstract createRelease(options: {
    owner: string;
    repo: string;
    tag: string;
    name: string;
    description: string;
    prerelease: boolean;
  }): Promise<void>;

  abstract getGitCredentials(): Promise<{
    email: string;
    username: string;
    password: string;
  }>;
}
