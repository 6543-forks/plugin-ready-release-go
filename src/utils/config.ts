import path from "path";
import { UserConfig } from "./types";

const ciConfig = {
  configFile: process.env.PLUGIN_CONFIG_FILE,
  isCI: process.env.CI === "woodpecker",
  eventType: process.env.PLUGIN_EVENT_TYPE || process.env.CI_PIPELINE_EVENT,
  branch: process.env.PLUGIN_BRANCH || process.env.CI_COMMIT_BRANCH,
  commitMessage:
    process.env.PLUGIN_COMMIT_MESSAGE || process.env.CI_COMMIT_MESSAGE,
  forgeType: process.env.PLUGIN_FORGE_TYPE || process.env.CI_REPO_FORGE_TYPE,
  githubToken: process.env.PLUGIN_GITHUB_TOKEN,
  gitEmail: process.env.PLUGIN_GIT_EMAIL,
  repoOwner: process.env.PLUGIN_REPO_OWNER || process.env.CI_REPO_OWNER,
  repoName: process.env.PLUGIN_REPO_NAME || process.env.CI_REPO_NAME,
};

export type Config = { user: UserConfig; ci: typeof ciConfig };

const defaultUserConfig: UserConfig = {
  getNextVersion: async () => "0.0.0", // TODO,
  getReleaseBranch: () => "main",
  getReleaseDescription: ({ nextVersion }) => `Release of ${nextVersion}`,
  getPullRequestBranch: ({ version }) => `next-release/${version}`,
  beforePrepare: async () => true,
  afterPrepare: async () => true,
  beforeRelease: async () => true,
  afterRelease: async () => true,
};

export async function getConfig(): Promise<Config> {
  const userConfig: Partial<UserConfig> = {};

  const configFilePath =
    ciConfig.configFile || path.join(process.cwd(), "release-config.ts");
  if (configFilePath) {
    console.log("# Loading config from", configFilePath, "...");
    const _userConfig = await import(configFilePath);
    Object.assign(userConfig, _userConfig.default);
    console.log("# Loaded config from", configFilePath);
  }

  return {
    user: { ...defaultUserConfig, ...userConfig },
    ci: ciConfig,
  };
}
