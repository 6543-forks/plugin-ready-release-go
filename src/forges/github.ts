import { Forge } from "./forge";
import { Octokit } from "@octokit/rest";

export class GithubForge extends Forge {
  accessToken: string;
  email: string;
  octokit: Octokit;

  constructor(accessToken: string, email: string) {
    super();
    this.email = email;
    this.accessToken = accessToken;
    this.octokit = new Octokit({
      auth: accessToken,
    });
  }

  async createOrUpdatePullRequest(options: {
    draft: boolean;
    title: string;
    description: string;
    owner: string;
    repo: string;
    sourceBranch: string;
    targetBranch: string;
  }): Promise<void> {
    const pullRequests = await this.octokit.pulls.list({
      owner: options.owner,
      repo: options.repo,
      head: `${options.owner}:${options.sourceBranch}`,
    });

    if (pullRequests.data.length > 0) {
      await this.octokit.pulls.update({
        owner: options.owner,
        repo: options.repo,
        pull_number: pullRequests.data[0].number,
        title: options.title,
        draft: options.draft,
        head: options.sourceBranch,
        base: options.targetBranch,
        body: options.description,
      });
      return;
    }

    await this.octokit.pulls.create({
      owner: options.owner,
      repo: options.repo,
      title: options.title,
      draft: options.draft,
      head: options.sourceBranch,
      base: options.targetBranch,
      body: options.description,
    });
  }

  async createRelease(options?: {
    owner: string;
    repo: string;
    tag: string;
    name: string;
    description: string;
    prerelease: boolean;
  }): Promise<void> {
    throw new Error("Method not implemented.");
  }

  async getGitCredentials(): Promise<{
    email: string;
    username: string;
    password: string;
  }> {
    return {
      email: this.email,
      username: "oauth",
      password: this.accessToken,
    };
  }
}
