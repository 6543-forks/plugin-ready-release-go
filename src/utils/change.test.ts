import { describe, it, expect } from "vitest";
import {
  getChangeLogSection,
  getNextVersionFromLabels,
  updateChangelogSection,
} from "./change";
import { Change } from "./types";
import { Config, defaultUserConfig } from "./config";
import { GithubForge } from "../forges/github";

const changes: Change[] = [
  {
    commitHash: "123",
    author: "John Doe",
    labels: ["bug", "ui"],
    title: "Fix random UI bug",
    pullRequestNumber: 1337,
  },
  {
    commitHash: "456",
    author: "Alice Wonderland",
    labels: ["feature"],
    title: "Add new feature",
    pullRequestNumber: 1338,
  },
  {
    commitHash: "789",
    author: "John Doe",
    labels: ["breaking"],
    title: "Break everything",
    pullRequestNumber: 1339,
  },
  {
    commitHash: "abc",
    author: "Alice Wonderland",
    labels: ["bug"],
    title: "Fix another bug",
  },
  {
    commitHash: "def",
    author: "Alice Wonderland",
    labels: [],
    title: "Update README",
  },
  {
    commitHash: "def",
    author: "Alice Wonderland",
    labels: ["enhancement"],
    title: "Update README",
  },
];
const config: Config = {
  ci: {
    branch: "main",
    commitMessage: "chore(release): 1.0.0 [skip ci]",
    configFile: "ready-go-release.config.ts",
    eventType: "push",
    gitEmail: "ci@woodpecker-ci.org",
    repoOwner: "woodpecker-ci",
    repoName: "woodpecker",
    forgeType: "github",
    githubToken: "123",
    isCI: true,
  },
  user: defaultUserConfig,
};

describe("change", () => {
  it("should get the correct major bump", () => {
    const nextVersion = getNextVersionFromLabels("1.0.0", config.user, changes);
    expect(nextVersion).toBe("2.0.0");
  });

  it("should get the correct minor bump", () => {
    const nextVersion = getNextVersionFromLabels(
      "1.0.0",
      config.user,
      changes.filter((change) => !change.labels.includes("breaking"))
    );

    expect(nextVersion).toBe("1.1.0");
  });

  it("should get the correct patch bump", () => {
    const nextVersion = getNextVersionFromLabels(
      "1.0.0",
      config.user,
      changes
        .filter((change) => !change.labels.includes("breaking"))
        .filter((change) => !change.labels.includes("feature"))
    );

    expect(nextVersion).toBe("1.0.1");
  });

  it("should generate a changelog", () => {
    const forge = new GithubForge("", "");
    const changelog = getChangeLogSection("1.0.0", config, changes, forge);

    expect(changelog).toMatchSnapshot();
  });

  it("should update changelog section", () => {
    const oldChangelog = `# Changelog

## [1.0.0](https://github.com/woodpecker-ci/woodpecker/releases/tag/1.0.0) - 2023-07-11

### ‚ú® Features

- Add new feature [#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338) ([456](https://github.com/woodpecker-ci/woodpecker/commit/456))

## [0.0.2](https://github.com/woodpecker-ci/woodpecker/releases/tag/0.0.1) - 2020-01-01

### üêõ Bug Fixes

- Fix random UI bug [#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337) ([123](https://github.com/woodpecker-ci/woodpecker/commit/123))
- Fix another bug [#43](https://github.com/woodpecker-ci/woodpecker/pull/43) ([456](https://github.com/woodpecker-ci/woodpecker/commit/456))

## [0.0.1](https://github.com/woodpecker-ci/woodpecker/releases/tag/0.0.1) - 2020-01-01

### ‚ú® Features

- Add new feature (#1338) @Alice Wonderland
    `;
    const forge = new GithubForge("", "");
    const nextVersion = "1.0.0";
    const newSection = getChangeLogSection(nextVersion, config, changes, forge);
    const changelog = updateChangelogSection(
      nextVersion,
      oldChangelog,
      newSection
    );

    expect(changelog).toMatchSnapshot();
  });
});
