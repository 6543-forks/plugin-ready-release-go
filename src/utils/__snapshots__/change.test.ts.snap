// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`change > should generate a changelog 1`] = `
"## [1.0.0](https://github.com/woodpecker-ci/woodpecker/releases/tag/1.0.0) - 2023-07-11

### 💥 Breaking changes

- Break everything [[#1339](https://github.com/woodpecker-ci/woodpecker/pull/1339)]

### ✨ Features

- Add new feature [[#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338)]

### 🐛 Bug Fixes

- Fix random UI bug [[#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337)]
- Fix another bug ([abc](https://github.com/woodpecker-ci/woodpecker/commit/abc))

### 📈 Enhancement

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

### Misc

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))"
`;

exports[`change > should update changelog section 1`] = `
"# Changelog

## [0.0.1](https://github.com/woodpecker-ci/woodpecker/releases/tag/0.0.1) - 2020-01-01

### ✨ Features

- Add new feature (#1338) @Alice Wonderland

## [0.0.2](https://github.com/woodpecker-ci/woodpecker/releases/tag/0.0.1) - 2020-01-01

### 🐛 Bug Fixes

- Fix random UI bug [#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337) ([123](https://github.com/woodpecker-ci/woodpecker/commit/123))
- Fix another bug [#43](https://github.com/woodpecker-ci/woodpecker/pull/43) ([456](https://github.com/woodpecker-ci/woodpecker/commit/456))

## [1.0.0](https://github.com/woodpecker-ci/woodpecker/releases/tag/1.0.0) - 2023-07-11

### 💥 Breaking changes

- Break everything [[#1339](https://github.com/woodpecker-ci/woodpecker/pull/1339)]

### ✨ Features

- Add new feature [[#1338](https://github.com/woodpecker-ci/woodpecker/pull/1338)]

### 🐛 Bug Fixes

- Fix random UI bug [[#1337](https://github.com/woodpecker-ci/woodpecker/pull/1337)]
- Fix another bug ([abc](https://github.com/woodpecker-ci/woodpecker/commit/abc))

### 📈 Enhancement

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))

### Misc

- Update README ([def](https://github.com/woodpecker-ci/woodpecker/commit/def))"
`;
